ambry-router/src/main/java/com.github.ambry.router/GCMCryptoService.java

```diff
@@ -93,6 +160,33 @@ public ByteBuffer decrypt(ByteBuffer toDecrypt, SecretKeySpec key) throws Genera
     }
   }
 
+  @Override
+  public ByteBuf decrypt(ByteBuf toDecrypt, SecretKeySpec key) throws GeneralSecurityException {
+    try {
+      Cipher decrypter = Cipher.getInstance(GCM_CRYPTO_INSTANCE, "BC");
```

Comment:
Same question here, since we are using heap buffers

Reply:
I take my words back about the nioBuffer. The implementation would actually save the memory when the buffer doesn't have array, since the temporary array it creates has a upper limit at 4096. If the buffer is larger than that, say 4MB, it will only create a 1/10 of what I did here, I will update.